File: solutions/2026-06-02/java/2026-06-02-earliest-finish-time-for-land-and-water-rides-i.java
```java
// Problem: Earliest Finish Time for Land and Water Rides I
// Link: https://leetcode.com/problems/earliest-finish-time-for-land-and-water-rides-i/
// Approach: Iterate through all possible combinations of one land ride and one water ride.
// For each pair, calculate the finish time if the land ride is taken first, and then if the water ride is taken first.
// The earliest of these finish times across all pairs will be the answer.
// Time Complexity: O(n*m), where n is the number of land rides and m is the number of water rides.
// This is because we have nested loops iterating through all land rides and water rides.
// Space Complexity: O(1), as we are only using a few variables to store intermediate results.

class Solution {
    public int earliestFinishTime(int[] landStartTime, int[] landDuration, int[] waterStartTime, int[] waterDuration) {
        // Initialize the minimum finish time to a very large value.
        int minFinishTime = Integer.MAX_VALUE;

        // Iterate through each land ride.
        for (int i = 0; i < landStartTime.length; i++) {
            // Iterate through each water ride.
            for (int j = 0; j < waterStartTime.length; j++) {

                // --- Scenario 1: Take land ride first, then water ride ---

                // Calculate the start time for the land ride. It's the maximum of its opening time
                // and the earliest possible boarding time (which is 0 initially if this is the first ride).
                int landRide1StartTime = landStartTime[i];
                // Calculate the finish time for the land ride.
                int landRide1FinishTime = landRide1StartTime + landDuration[i];

                // Calculate the start time for the water ride. It's the maximum of its opening time
                // and the finish time of the land ride.
                int waterRide1StartTime = Math.max(waterStartTime[j], landRide1FinishTime);
                // Calculate the finish time for the water ride.
                int waterRide1FinishTime = waterRide1StartTime + waterDuration[j];

                // Update the minimum finish time if this scenario yields an earlier finish time.
                minFinishTime = Math.min(minFinishTime, waterRide1FinishTime);

                // --- Scenario 2: Take water ride first, then land ride ---

                // Calculate the start time for the water ride. It's the maximum of its opening time
                // and the earliest possible boarding time (which is 0 initially if this is the first ride).
                int waterRide2StartTime = waterStartTime[j];
                // Calculate the finish time for the water ride.
                int waterRide2FinishTime = waterRide2StartTime + waterDuration[j];

                // Calculate the start time for the land ride. It's the maximum of its opening time
                // and the finish time of the water ride.
                int landRide2StartTime = Math.max(landStartTime[i], waterRide2FinishTime);
                // Calculate the finish time for the land ride.
                int landRide2FinishTime = landRide2StartTime + landDuration[i];

                // Update the minimum finish time if this scenario yields an earlier finish time.
                minFinishTime = Math.min(minFinishTime, landRide2FinishTime);
            }
        }

        // Return the overall earliest finish time found.
        return minFinishTime;
    }
}
```

File: solutions/2026-06-02/cpp/2026-06-02-earliest-finish-time-for-land-and-water-rides-i.cpp
```cpp
// Solves LeetCode problem "Earliest Finish Time for Land and Water Rides I".
// The problem asks for the minimum time to complete one land ride and one water ride.
// We can take the rides in any order, and must respect their start times and durations.
//
// Approach:
// The core idea is to iterate through all possible combinations of one land ride and one water ride.
// For each pair of rides (one land, one water), we calculate the earliest finish time if the land ride is taken first,
// and then the earliest finish time if the water ride is taken first.
// We keep track of the minimum finish time found across all pairs and all orders.
//
// Let's consider a land ride `i` and a water ride `j`.
//
// Case 1: Land ride `i` first, then water ride `j`.
//   - Start land ride `i` at `landStartTime[i]`.
//   - Finish land ride `i` at `landFinishTime = landStartTime[i] + landDuration[i]`.
//   - The water ride `j` can be started at the maximum of its start time (`waterStartTime[j]`) and when the land ride finishes (`landFinishTime`).
//   - So, `waterStartTimeActual = max(waterStartTime[j], landFinishTime)`.
//   - Finish water ride `j` at `waterFinishTime = waterStartTimeActual + waterDuration[j]`.
//   - The total finish time for this sequence is `waterFinishTime`.
//
// Case 2: Water ride `j` first, then land ride `i`.
//   - Start water ride `j` at `waterStartTime[j]`.
//   - Finish water ride `j` at `waterFinishTime = waterStartTime[j] + waterDuration[j]`.
//   - The land ride `i` can be started at the maximum of its start time (`landStartTime[i]`) and when the water ride finishes (`waterFinishTime`).
//   - So, `landStartTimeActual = max(landStartTime[i], waterFinishTime)`.
//   - Finish land ride `i` at `landFinishTime = landStartTimeActual + landDuration[i]`.
//   - The total finish time for this sequence is `landFinishTime`.
//
// We initialize `minFinishTime` to a very large value. For each pair of rides, we calculate the finish times for both orders
// and update `minFinishTime` if a smaller finish time is found.
//
// Time Complexity:
// Let n be the number of land rides and m be the number of water rides.
// We have a nested loop iterating through all n land rides and all m water rides.
// Inside the loop, we perform constant-time operations (max, addition).
// Therefore, the time complexity is O(n * m).
//
// Space Complexity:
// We are only using a few variables to store the minimum finish time and loop indices.
// The space used does not depend on the input size.
// Therefore, the space complexity is O(1).
//
// Link: https://leetcode.com/problems/earliest-finish-time-for-land-and-water-rides-i/

#include <vector>
#include <algorithm> // For std::max
#include <limits>    // For std::numeric_limits

class Solution {
public:
    int earliestFinishTime(std::vector<int>& landStartTime, std::vector<int>& landDuration, std::vector<int>& waterStartTime, std::vector<int>& waterDuration) {
        // Initialize the minimum finish time to the largest possible integer value.
        // This ensures that any valid finish time will be smaller and will update this variable.
        int minFinishTime = std::numeric_limits<int>::max();

        // Iterate through each land ride.
        for (int i = 0; i < landStartTime.size(); ++i) {
            // Iterate through each water ride.
            for (int j = 0; j < waterStartTime.size(); ++j) {

                // --- Scenario 1: Land ride first, then Water ride ---

                // Calculate the finish time of the current land ride.
                // The land ride starts at its earliest possible start time.
                int landFinishTime1 = landStartTime[i] + landDuration[i];

                // The water ride can start only after it opens AND after the land ride finishes.
                // So, the actual start time for the water ride is the maximum of its opening time and the land ride's finish time.
                int waterStartTimeActual1 = std::max(waterStartTime[j], landFinishTime1);

                // Calculate the finish time of the water ride in this scenario.
                int waterFinishTime1 = waterStartTimeActual1 + waterDuration[j];

                // Update the overall minimum finish time if this scenario yields an earlier finish.
                minFinishTime = std::min(minFinishTime, waterFinishTime1);

                // --- Scenario 2: Water ride first, then Land ride ---

                // Calculate the finish time of the current water ride.
                // The water ride starts at its earliest possible start time.
                int waterFinishTime2 = waterStartTime[j] + waterDuration[j];

                // The land ride can start only after it opens AND after the water ride finishes.
                // So, the actual start time for the land ride is the maximum of its opening time and the water ride's finish time.
                int landStartTimeActual2 = std::max(landStartTime[i], waterFinishTime2);

                // Calculate the finish time of the land ride in this scenario.
                int landFinishTime2 = landStartTimeActual2 + landDuration[i];

                // Update the overall minimum finish time if this scenario yields an earlier finish.
                minFinishTime = std::min(minFinishTime, landFinishTime2);
            }
        }

        // Return the earliest possible finish time found among all combinations and orders.
        return minFinishTime;
    }
};
```